Basket-Donation-Payments/Basket-Donation-Payments/View Controllers/Login & Signup Flow/NonprofitSignupViewController.m

```diff
@@ -61,31 +61,66 @@ -(void)saveNonprofitAndUserToParse {
     self.user.nonprofit = self.nonprofit;
-    [self.user signUpInBackgroundWithBlock:^(BOOL succeeded, NSError * _Nullable error) {
-        if (succeeded) {
-            if (self.finishedSavingBoth) {
-                [self performSegueWithIdentifier:@"loginSegue" sender:nil];
-            } else {
-                self.finishedSavingBoth = true;
-            }
+    
+    NSOperationQueue *opQueue = [NSOperationQueue new];
 
-        } else {
-            UIAlertController *alert = [Utils createAlertControllerWithTitle:@"Could not save user." andMessage:error.localizedDescription okCompletion:nil cancelCompletion:nil];
-            [self presentViewController:alert animated:YES completion:nil];
-        }
+    
+    NSBlockOperation *signupUserOp = [NSBlockOperation new];
+    signupUserOp.queuePriority = NSOperationQueuePriorityHigh;
+    __weak NSBlockOperation *weakSignupUserOp = signupUserOp;
+    [signupUserOp addExecutionBlock:^{
+        [self.user signUpInBackgroundWithBlock:^(BOOL succeeded, NSError * _Nullable error) {
+            if (!succeeded) {
+                [opQueue cancelAllOperations];
+                UIAlertController *alert = [Utils createAlertControllerWithTitle:@"Could not save user." andMessage:error.localizedDescription okCompletion:nil cancelCompletion:nil];
+                [self presentViewController:alert animated:YES completion:nil];
+            }
+        }];
     }];
-    [self.nonprofit saveInBackgroundWithBlock:^(BOOL succeeded, NSError * _Nullable error) {
-        if (succeeded) {
-            if (self.finishedSavingBoth) {
-                [self performSegueWithIdentifier:@"loginSegue" sender:nil];
-            } else {
-                self.finishedSavingBoth = true;
+    
+    NSBlockOperation *saveNonprofitOp = [NSBlockOperation new];
+    saveNonprofitOp.queuePriority = NSOperationQueuePriorityHigh;
+    __weak NSBlockOperation *weakSaveNonprofitOp = saveNonprofitOp;
+    [saveNonprofitOp addExecutionBlock:^{
+        [self.user signUpInBackgroundWithBlock:^(BOOL succeeded, NSError * _Nullable error) {
+            if ([weakSaveNonprofitOp isCancelled]) {
+                [self.user deleteInBackgroundWithBlock:^(BOOL succeeded, NSError * _Nullable error) {
+                    if (!succeeded) {
+                        NSLog(@"error");
+                    }
+                }];
+                return;
+            } else if (!succeeded) {
+                [opQueue cancelAllOperations];
+                UIAlertController *alert = [Utils createAlertControllerWithTitle:@"Could not save nonprofit." andMessage:error.localizedDescription okCompletion:nil cancelCompletion:nil];
+                [self presentViewController:alert animated:YES completion:nil];
             }
+        }];
+    }];
+    
+    
+    NSBlockOperation *loginSegueOp = [NSBlockOperation new];
+    //FIXME: loginSegueOp is still running first??
+    loginSegueOp.queuePriority = NSOperationQueuePriorityLow;
```

Comment:
Actually you shouldn't use a regular operation for the login segue, since you want this to be executed after all operations are complete. A lower priority operation might work but it's kind of hacky. You should use `waitUntilAllOperationsAreFinished` on the `NSOperationQueue` instead, but you have to use a background thread to avoid blocking the main thread.